Adding NMS helm chart for Magma

reddydodda opened the pull request:
This will help to deploy NMS on K8S using helm chart

Signed-off-by: reddydodda <reddydodda@gmail.com>

reddydodda committed:
fix repo path

reddydodda committed:
Adding NMS helm chart for Magma

This will help to deploy NMS on K8S using helm chart

Signed-off-by: reddydodda <reddydodda@gmail.com>

rpraveen commented on nms/helmchart/nms/Chart.yaml:
```diff
@@ -0,0 +1,27 @@
+  - nms
+  - orc8r
+maintainers:
+  - name: Mirantis Inc
```
May I suggest removing the maintainers, considering the helm chart will be part of the https://github.com/facebookincubator/magma open source?

reddydodda replied:
fixed

rpraveen commented on nms/helmchart/nms/templates/magmalte-deployment.yaml:
```diff
@@ -0,0 +1,142 @@
+            - containerPort: 8081
+          volumeMounts:
+            - name: orc8r-secrets-certs
+              mountPath: /usr/src/fbcnms-projects/magmalte/certs/admin_operator.pem
```
Please add these 2 file names to the "Copy orchestrator secrets" section in https://github.com/facebookincubator/magma/blob/master/orc8r/cloud/helm/orc8r/README.md

rpraveen commented on nms/helmchart/nms/README.md:
```diff
@@ -0,0 +1,113 @@
+## NMS
+
+Install NMS which includes nginx and magmalte subcomponents.
```
@reddydodda what do you think about moving this chart to https://github.com/facebookincubator/magma/tree/master/orc8r/cloud/helm/orc8r/charts, so that we have all the helm related content for orc8r in one place?

Also it would be very useful to point to this README in https://github.com/facebookincubator/magma/blob/master/orc8r/cloud/helm/orc8r/README.md, as another step.

reddydodda replied:
fixed

reddydodda replied:
@rpraveen  so regarding the second point. how about moving all helm charts to new folder ? in that case we can have a separate installation doc for helm charts. next, I will push charts for AGW and FeG.  having all at the same location makes sense. what do you think?

rpraveen replied:
@reddydodda good point. I was thinking that orc8r as a whole needs one helm folder. FeG and AGW are independent components, and for fixed wireless deployments one doesnt need to run FeG. Shall we move this to orc8r/cloud/helm/orc8r/charts, and then I can pull this PR?

reddydodda replied:
okay .. will do that

rpraveen commented on nms/helmchart/nms/templates/nginx-proxy-deployment.yaml:
```diff
@@ -0,0 +1,164 @@
+              mountPath: /etc/nginx/conf.d
+            - name: orc8r-secrets-certs
+              mountPath: /etc/nginx/conf.d/nms_nginx.pem
+              subPath: nms_nginx.pem
```
Lets add these 2 filenames to the "Copy orchestrator secrets" section in the README too.

rpraveen commented on nms/helmchart/nms/README.md:
```diff
@@ -0,0 +1,113 @@
+| `service.http.targetport` | Service targetport number for magmalte | `8081` |
+| `service.http.nodePort` | Service nodePort number | `""` |
+| Nginx:       |
+| `manifests.configmap` | Enable Nginx configmap to store config files. | `true` |
```
We dont seem to reference nginx.manifests.configmap and nginx.manifests.secrets anywhere?

reddydodda replied:
fixed

rpraveen commented on nms/helmchart/nms/README.md:
```diff
@@ -0,0 +1,113 @@
+$ cat vals.yaml
+magmalte:
+  env:
+    api_host: orc8r-proxy.magma.svc.cluster.local:9443
```
Just to keep the vals.yaml file minimal, we wouldnt need to change api_host, host and port right? Shall we remove them from here?

rpraveen commented on nms/helmchart/nms/README.md:
````diff
@@ -0,0 +1,113 @@
+$ helm upgrade --install nms ./nms --namespace=magma
+
+```
+- Create test user:
````
s/Create test user/Create admin user/ :)

rpraveen commented on nms/helmchart/nms/templates/helpers/template.tpl:
```diff
@@ -0,0 +1,23 @@
+{{- $name := index . 0 -}}
+{{- $context := index . 1 -}}
+{{- $last := base $context.Template.Name }}
+{{- $wtf := $context.Template.Name | replace $last $name -}}
```
Can we change the variable name to '$filepath' or something clearer? :)

rpraveen reviewed (changes requested):
Nice! Looks great!

reddydodda committed:
Fix LICENSE, doc and Serviceaccount for NMS chart

Signed-off-by: reddydodda <reddydodda@gmail.com>

reddydodda committed:
Move NMS chart to orc8s sub-chart and diable the NMS subchart by default

Signed-off-by: reddydodda <reddydodda@gmail.com>

reddydodda committed:
fix doc

reddydodda committed:
delete nms helm chart from nms dir

reddydodda committed:
disable nms by default and fix rbac

reddydodda committed:
fix private registry image pull logic

xjtian commented on orc8r/cloud/helm/orc8r/charts/nms/templates/helpers/pod_anti_affinity.tpl:
```diff
@@ -0,0 +1,43 @@
+{{/*
```
why do we need this template?

reddydodda replied:
@xjtian  this is required for pod advanced scheduling . in some cases if we need to scale pod-a on the same node where already that pod is running then this will help, also vice-versa.

xjtian replied:
same as the resource template and the service accounts, let's add this when we need it, to the helpers toolbox. at the scale of our typical upcoming deployments this seems unnecessary.

xjtian commented on orc8r/cloud/helm/orc8r/charts/nms/templates/nginx-proxy-deployment.yaml:
```diff
@@ -0,0 +1,166 @@
+  name: {{ $serviceAccountName }}
+  apiGroup: rbac.authorization.k8s.io
+{{- end }}
+---
```
why do we create service accounts for each separate nms deployment?

reddydodda replied:
@xjtian the motive is to have different Roles for magmalte and nginx. if you don't see any use of it, then I can move it to have one for both pods.

xjtian replied:
yeah I think this is overkill. can we just drop the role entirely? we don't set this up for the other deployments either

xjtian commented on orc8r/cloud/helm/orc8r/charts/nms/templates/helpers/resources.tpl:
```diff
@@ -0,0 +1,34 @@
+{{/*
```
if we see a need for something like this, this is something we'd probably want to pull into the base orc8r chart as a helper which can be used by nms/metrics charts. but until we hit something like that, IMO we should KISS and leave it out for now

reddydodda replied:
yes ... the main goal is to have a common helper file for all charts. so in my next commit, I will push updates to orc8r and metric chart to have a single helper and has a reference from there to the helper. we can name it something like magma-helper-toolkit. we can leverage this going forward.  

I agree this is too early but keeping in mind magma roadmap I propose to go with this way.

xjtian replied:
If we don't have an identified need, we should drop this for now and avoid the extra complexity. It's pretty unlikely that NMS would hog node resources in any significant way. The toolkit is a good idea, but one we should flesh out later

xjtian reviewed (commented):
mainly just some questions here for clarity. There's a lot more going on here than I expected, hoping to clear up some of the complexity

reddydodda committed:
update deployment tot have one service account for both magalte and nginx

xjtian reviewed (approved):
LGTM now, but can you gate the subchart so it doesn't get deployed by default? We'll need to update some of our deployment tooling on our end so this is deployable, and I don't want oncalls borking the weekly release before that happens :)

I think we do something similar for the secrets subchart.

reddydodda commented on orc8r/cloud/helm/orc8r/values.yaml:
```diff
@@ -5,11 +5,30 @@
+nms:
+  magmalte:
+    manifests:
+      secrets: false
```
@xjtian  default NMS is disabled.

xjtian commented:
This looks ready to go @reddydodda, but can you rebase this for me on top of head master? Looks like there are some merge conflicts with the helm readme

reddydodda committed:
This commit fixs

1. README file
2. helper file when NMS is used as subchart
3. livenessProbe for NMS

Signed-off-by: reddydodda <reddydodda@gmail.com>

reddydodda committed:
fix label reference for NMS chart
